File: benchmarks/handwritten_baselines/authoring_density/enterprise_monitor.py
"""
Authoring-density baseline: enterprise infrastructure health monitor — idiomatic Python.

Semantically equivalent to examples/benchmark/enterprise_monitor.ainl.
Written in the style a proficient Python developer (or LLM) would produce
when asked to "build a health monitor that polls an HTTP endpoint, routes
by severity, generates LLM incident alerts only when needed, and caches state."

Dependencies: httpx, openai  (pip install httpx openai)
"""

from __future__ import annotations

import json
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Literal, Optional

import httpx
from openai import OpenAI


Severity = Literal["healthy", "degraded", "critical"]


# ---------------------------------------------------------------------------
# Simple file-backed cache (mirrors cache adapter behaviour)
# ---------------------------------------------------------------------------

class FileCache:
    """Persistent JSON key-value store backed by a local file."""

    def __init__(self, path: str = ".monitor_cache.json") -> None:
        self._path = Path(path)
        self._data: dict[str, str] = {}
        if self._path.exists():
            try:
                self._data = json.loads(self._path.read_text())
            except (json.JSONDecodeError, OSError):
                self._data = {}

    def get(self, key: str) -> Optional[str]:
        return self._data.get(key)

    def set(self, key: str, value: str) -> None:
        self._data[key] = value
        self._path.write_text(json.dumps(self._data, indent=2))


# ---------------------------------------------------------------------------
# Result type
# ---------------------------------------------------------------------------

@dataclass
class MonitorResult:
    severity: Severity
    endpoint_url: str
    status_code: Optional[int]
    latency_ms: Optional[float]
    alert_text: Optional[str]


# ---------------------------------------------------------------------------
# Pipeline
# ---------------------------------------------------------------------------

def check_endpoint(
    endpoint_url: str,
    threshold_ms: float,
    cache: Optional[FileCache] = None,
    openai_client: Optional[OpenAI] = None,
    http_timeout: float = 10.0,
) -> MonitorResult:
    """
    Poll endpoint, evaluate status + latency, route by severity, generate
    LLM alert only when degraded or down, cache severity state.
    """
    if cache is None:
        cache = FileCache()
    if openai_client is None:
        openai_client = OpenAI()

    # Poll the health endpoint
    t0 = time.monotonic()
    try:
        with httpx.Client(timeout=http_timeout) as client:
            response = client.get(endpoint_url)
        latency_ms = (time.monotonic() - t0) * 1000.0
        status_code: Optional[int] = response.status_code
        is_up = (status_code == 200)
    except (httpx.ConnectError, httpx.TimeoutException):
        latency_ms = (time.monotonic() - t0) * 1000.0
        status_code = None
        is_up = False

    # Routing — zero LLM tokens
    if not is_up:
        severity: Severity = "critical"
    elif latency_ms > threshold_ms:
        severity = "degraded"
    else:
        severity = "healthy"

    # LLM alert generation — only for non-healthy states
    alert_text: Optional[str] = None

    if severity == "critical":
        prompt = (
            f"Critical: endpoint {endpoint_url} is DOWN. "
            f"HTTP status: {status_code}. "
            "Draft a concise ops incident alert."
        )
        completion = openai_client.chat.completions.create(
            model="gpt-4o",
            messages=[{"role": "user", "content": prompt}],
            max_tokens=200,
            temperature=0.2,
        )
        alert_text = completion.choices[0].message.content or ""

    elif severity == "degraded":
        prompt = (
            f"Degraded: endpoint {endpoint_url} latency {latency_ms:.0f}ms "
            f"exceeds threshold {threshold_ms:.0f}ms. "
            "Draft a concise ops alert."
        )
        completion = openai_client.chat.completions.create(
            model="gpt-4o",
            messages=[{"role": "user", "content": prompt}],
            max_tokens=150,
            temperature=0.3,
        )
        alert_text = completion.choices[0].message.content or ""

    # Cache severity state
    cache.set("monitor_last_severity", severity)
    if alert_text is not None:
        cache.set("monitor_last_alert", alert_text)

    return MonitorResult(
        severity=severity,
        endpoint_url=endpoint_url,
        status_code=status_code,
        latency_ms=latency_ms,
        alert_text=alert_text,
    )


# ---------------------------------------------------------------------------
# CLI entry point
# ---------------------------------------------------------------------------

if __name__ == "__main__":
    import sys

    endpoint_url = sys.argv[1] if len(sys.argv) > 1 else "https://httpbin.org/status/200"
    threshold_ms = float(sys.argv[2]) if len(sys.argv) > 2 else 500.0
    result = check_endpoint(endpoint_url, threshold_ms)
    print(json.dumps(vars(result), indent=2))

File: benchmarks/handwritten_baselines/authoring_density/enterprise_monitor.ts
/**
 * Authoring-density baseline: enterprise health monitor — idiomatic TypeScript.
 *
 * Semantically equivalent to examples/benchmark/enterprise_monitor.ainl.
 * Written in the style a proficient TypeScript developer (or LLM) would produce
 * when asked to "build a health monitor that polls an HTTP endpoint, routes
 * by severity, generates LLM incident alerts only when needed, and caches state."
 *
 * Dependencies: openai  (npm install openai)
 */

import OpenAI from "openai";
import { readFileSync, writeFileSync, existsSync } from "fs";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export type Severity = "healthy" | "degraded" | "critical";

export interface MonitorResult {
  severity: Severity;
  endpointUrl: string;
  statusCode: number | null;
  latencyMs: number;
  alertText: string | null;
}

// ---------------------------------------------------------------------------
// File-backed cache (mirrors AINL cache adapter behaviour)
// ---------------------------------------------------------------------------

class FileCache {
  private readonly path: string;
  private data: Record<string, string>;

  constructor(path = ".monitor_cache.json") {
    this.path = path;
    try {
      this.data = existsSync(path)
        ? (JSON.parse(readFileSync(path, "utf8")) as Record<string, string>)
        : {};
    } catch {
      this.data = {};
    }
  }

  set(key: string, value: string): void {
    this.data[key] = value;
    writeFileSync(this.path, JSON.stringify(this.data, null, 2));
  }
}

// ---------------------------------------------------------------------------
// Pipeline
// ---------------------------------------------------------------------------

export async function checkEndpoint(
  endpointUrl: string,
  thresholdMs: number,
  options?: {
    cache?: FileCache;
    openai?: OpenAI;
    httpTimeoutMs?: number;
  }
): Promise<MonitorResult> {
  const cache = options?.cache ?? new FileCache();
  const openai = options?.openai ?? new OpenAI();
  const timeoutMs = options?.httpTimeoutMs ?? 10_000;

  // Poll the health endpoint
  const t0 = Date.now();
  let statusCode: number | null = null;
  let isUp = false;

  try {
    const response = await fetch(endpointUrl, {
      signal: AbortSignal.timeout(timeoutMs),
    });
    statusCode = response.status;
    isUp = statusCode === 200;
  } catch {
    // Network error / timeout — treat as down
  }
  const latencyMs = Date.now() - t0;

  // Severity routing — zero LLM tokens
  let severity: Severity;
  if (!isUp) {
    severity = "critical";
  } else if (latencyMs > thresholdMs) {
    severity = "degraded";
  } else {
    severity = "healthy";
  }

  // LLM alert generation — only for non-healthy states
  let alertText: string | null = null;

  if (severity === "critical") {
    const prompt =
      `Critical: endpoint ${endpointUrl} is DOWN. ` +
      `HTTP status: ${statusCode ?? "N/A"}. ` +
      "Draft a concise ops incident alert.";
    const completion = await openai.chat.completions.create({
      model: "gpt-4o",
      messages: [{ role: "user", content: prompt }],
      max_tokens: 200,
      temperature: 0.2,
    });
    alertText = completion.choices[0]?.message.content ?? null;
  } else if (severity === "degraded") {
    const prompt =
      `Degraded: endpoint ${endpointUrl} latency ${latencyMs}ms ` +
      `exceeds threshold ${thresholdMs}ms. ` +
      "Draft a concise ops alert.";
    const completion = await openai.chat.completions.create({
      model: "gpt-4o",
      messages: [{ role: "user", content: prompt }],
      max_tokens: 150,
      temperature: 0.3,
    });
    alertText = completion.choices[0]?.message.content ?? null;
  }

  // Cache severity state
  cache.set("monitor_last_severity", severity);
  if (alertText !== null) {
    cache.set("monitor_last_alert", alertText);
  }

  return { severity, endpointUrl, statusCode, latencyMs, alertText };
}

// ---------------------------------------------------------------------------
// CLI entry point
// ---------------------------------------------------------------------------

if (process.argv[1] === new URL(import.meta.url).pathname) {
  const endpointUrl = process.argv[2] ?? "https://httpbin.org/status/200";
  const thresholdMs = Number(process.argv[3] ?? 500);

  checkEndpoint(endpointUrl, thresholdMs)
    .then((r) => console.log(JSON.stringify(r, null, 2)))
    .catch((e) => {
      console.error(e);
      process.exit(1);
    });
}
